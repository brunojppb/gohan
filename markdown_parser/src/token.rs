use std::fmt::{Debug, Display};

use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize)]
pub struct Token {
    pub position: usize,
    pub kind: TokenType,
    pub line: usize,
}

impl Token {
    pub fn new(position: usize, kind: TokenType, line: usize) -> Self {
        Self {
            position,
            kind,
            line,
        }
    }
}

#[derive(Debug, Serialize, Deserialize)]
pub enum TokenType {
    Hash,
    Star,
    Bang,
    Underscore,
    Newline,
    Tab,
    Space,
    Dot,
    Dash,
    Backslash,
    LeftParen,
    RightParen,
    LeftSquareBraket,
    RightSquareBraket,
    Number(usize),
    Text(String),
    EndOfFile,
}

impl TokenType {
    /// Lexeme representation. Mostly for debugging
    pub fn lexime(&self) -> String {
        match self {
            Self::Hash => "#".to_string(),
            Self::Star => "*".to_string(),
            Self::Bang => "!".to_string(),
            Self::Dot => ".".to_string(),
            Self::Dash => "-".to_string(),
            Self::Backslash => "\\".to_string(),
            Self::LeftParen => "(".to_string(),
            Self::RightParen => ")".to_string(),
            Self::LeftSquareBraket => "[".to_string(),
            Self::RightSquareBraket => "]".to_string(),
            Self::Tab => "\\t".to_string(),
            Self::Space => " ".to_string(),
            Self::Newline => "\\n".to_string(),
            Self::Underscore => "_".to_string(),
            Self::Number(number) => number.to_string(),
            Self::Text(text) => text.to_string(),
            Self::EndOfFile => "EOF".to_string(),
        }
    }
}

impl Display for TokenType {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "lexeme={}", self.lexime())
    }
}
