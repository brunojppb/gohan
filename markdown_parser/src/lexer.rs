use crate::token::{Token, TokenType};

const SPECIAL_TOKENS: &str = "#*!_[]().- \n\t\\";

pub struct Lexer<'a> {
    input: &'a str,
    tokens: Vec<Token>,
    start: usize,
    current: usize,
    line: usize,
}

impl<'a> Lexer<'a> {
    pub fn new(input: &'a str) -> Self {
        Self {
            input,
            tokens: Vec::new(),
            start: 0,
            current: 0,
            line: 1,
        }
    }

    pub fn scan(&mut self) -> &Vec<Token> {
        while !self.is_at_end() {
            self.start = self.current;
            self.scan_token();
        }

        self.add_token(TokenType::EndOfFile);
        &self.tokens
    }

    fn scan_token(&mut self) {
        let Some(c) = self.advance() else {
            panic!("Could not scan the next token. Line {}", &self.line);
        };

        match c {
            '#' => self.add_token(TokenType::Hash),
            '*' => self.add_token(TokenType::Star),
            '!' => self.add_token(TokenType::Bang),
            ' ' => self.add_token(TokenType::Space),
            '_' => self.add_token(TokenType::Underscore),
            '-' => self.add_token(TokenType::Dash),
            '.' => self.add_token(TokenType::Dot),
            '(' => self.add_token(TokenType::LeftParen),
            ')' => self.add_token(TokenType::RightParen),
            '[' => self.add_token(TokenType::LeftSquareBraket),
            ']' => self.add_token(TokenType::RightSquareBraket),
            '\\' => self.add_token(TokenType::Backslash),
            '\t' => self.add_token(TokenType::Tab),
            '\n' => {
                self.line += 1;
                self.add_token(TokenType::Newline);
            }
            c if c.is_ascii_digit() => {
                self.add_token(TokenType::Number(c.to_digit(10).unwrap() as usize))
            }
            _ => self.handle_string(),
        }
    }

    fn is_token(&self, c: Option<char>) -> bool {
        let r = c
            .filter(|c| c.is_ascii_digit() || SPECIAL_TOKENS.contains(*c))
            .is_some();
        println!("c={:?} r={}", c, r);
        r
    }

    fn handle_string(&mut self) {
        while !self.is_at_end() && !self.is_token(self.peek()) {
            self.advance();
        }

        let value: String = self
            .input
            .chars()
            .skip(self.start)
            .take(self.current - self.start)
            .collect();

        println!(
            "start={} current={} value='{}'",
            self.start, self.current, value
        );

        self.add_token(TokenType::Text(value));
    }

    fn is_at_end(&self) -> bool {
        self.current >= self.input.len()
    }

    fn add_token(&mut self, token_type: TokenType) {
        let token = Token::new(self.start, token_type, self.line);

        self.tokens.push(token);
    }

    /// Look-up the next character, but do not consume it
    fn peek(&self) -> Option<char> {
        if self.is_at_end() {
            return None;
        }
        self.input.chars().nth(self.current)
    }

    /// Consume the next character and advance the needle
    /// to point to a potential next character
    fn advance(&mut self) -> Option<char> {
        let c = self.input.chars().nth(self.current);
        self.current += 1;
        c
    }

    // Look-up one character after the next, but do not consume it
    // fn peek_next(&self) -> Option<char> {
    //     self.input.chars().nth(self.current + 1)
    // }
    //
    // fn match_next(&mut self, expected: char) -> bool {
    //     if self.is_at_end() {
    //         return false;
    //     }

    //     if self.peek() == Some(expected) {
    //         self.advance();
    //         return true;
    //     }

    //     false
    // }
}

#[cfg(test)]
mod tests {
    use std::fs;

    use super::*;

    #[test]
    fn parse_headers() {
        insta::glob!("snapshot_inputs/*.md", |path| {
            let markdown = fs::read_to_string(path).unwrap();
            let mut lexer = Lexer::new(&markdown);
            insta::assert_json_snapshot!(lexer.scan());
        });
    }
}
