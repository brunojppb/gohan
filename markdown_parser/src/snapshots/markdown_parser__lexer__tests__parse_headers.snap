---
source: markdown_parser/src/lexer.rs
expression: lexer.scan()
input_file: markdown_parser/src/snapshot_inputs/lexer_input.md
---
[
  {
    "position": 0,
    "kind": "Hash",
    "line": 1
  },
  {
    "position": 1,
    "kind": "Space",
    "line": 1
  },
  {
    "position": 2,
    "kind": {
      "Text": "Hi"
    },
    "line": 1
  },
  {
    "position": 4,
    "kind": "Space",
    "line": 1
  },
  {
    "position": 5,
    "kind": {
      "Text": "there"
    },
    "line": 1
  },
  {
    "position": 10,
    "kind": "Newline",
    "line": 2
  },
  {
    "position": 11,
    "kind": "Newline",
    "line": 3
  },
  {
    "position": 12,
    "kind": {
      "Text": "This"
    },
    "line": 3
  },
  {
    "position": 16,
    "kind": "Space",
    "line": 3
  },
  {
    "position": 17,
    "kind": {
      "Text": "should"
    },
    "line": 3
  },
  {
    "position": 23,
    "kind": "Space",
    "line": 3
  },
  {
    "position": 24,
    "kind": {
      "Text": "be"
    },
    "line": 3
  },
  {
    "position": 26,
    "kind": "Space",
    "line": 3
  },
  {
    "position": 27,
    "kind": {
      "Text": "text"
    },
    "line": 3
  },
  {
    "position": 31,
    "kind": "Space",
    "line": 3
  },
  {
    "position": 32,
    "kind": "Underscore",
    "line": 3
  },
  {
    "position": 33,
    "kind": {
      "Text": "wrapped"
    },
    "line": 3
  },
  {
    "position": 40,
    "kind": "Space",
    "line": 3
  },
  {
    "position": 41,
    "kind": {
      "Text": "in"
    },
    "line": 3
  },
  {
    "position": 43,
    "kind": "Space",
    "line": 3
  },
  {
    "position": 44,
    "kind": {
      "Text": "underscore"
    },
    "line": 3
  },
  {
    "position": 54,
    "kind": "Underscore",
    "line": 3
  },
  {
    "position": 55,
    "kind": "Space",
    "line": 3
  },
  {
    "position": 56,
    "kind": {
      "Text": "and"
    },
    "line": 3
  },
  {
    "position": 59,
    "kind": "Space",
    "line": 3
  },
  {
    "position": 60,
    "kind": "Backslash",
    "line": 3
  },
  {
    "position": 61,
    "kind": "Backslash",
    "line": 3
  },
  {
    "position": 62,
    "kind": {
      "Text": "escaped"
    },
    "line": 3
  },
  {
    "position": 69,
    "kind": "Newline",
    "line": 4
  },
  {
    "position": 70,
    "kind": "Star",
    "line": 4
  },
  {
    "position": 71,
    "kind": "Star",
    "line": 4
  },
  {
    "position": 72,
    "kind": {
      "Text": "bold"
    },
    "line": 4
  },
  {
    "position": 76,
    "kind": "Space",
    "line": 4
  },
  {
    "position": 77,
    "kind": {
      "Text": "text"
    },
    "line": 4
  },
  {
    "position": 81,
    "kind": "Space",
    "line": 4
  },
  {
    "position": 82,
    "kind": {
      "Text": "here"
    },
    "line": 4
  },
  {
    "position": 86,
    "kind": "Star",
    "line": 4
  },
  {
    "position": 87,
    "kind": "Star",
    "line": 4
  },
  {
    "position": 88,
    "kind": "Space",
    "line": 4
  },
  {
    "position": 89,
    "kind": {
      "Text": "with"
    },
    "line": 4
  },
  {
    "position": 93,
    "kind": "Space",
    "line": 4
  },
  {
    "position": 94,
    "kind": {
      "Text": "a"
    },
    "line": 4
  },
  {
    "position": 95,
    "kind": "Space",
    "line": 4
  },
  {
    "position": 96,
    "kind": {
      "Text": "nice"
    },
    "line": 4
  },
  {
    "position": 100,
    "kind": "Space",
    "line": 4
  },
  {
    "position": 101,
    "kind": "LeftSquareBraket",
    "line": 4
  },
  {
    "position": 102,
    "kind": {
      "Text": "link"
    },
    "line": 4
  },
  {
    "position": 106,
    "kind": "Space",
    "line": 4
  },
  {
    "position": 107,
    "kind": {
      "Text": "here"
    },
    "line": 4
  },
  {
    "position": 111,
    "kind": "RightSquareBraket",
    "line": 4
  },
  {
    "position": 112,
    "kind": "LeftParen",
    "line": 4
  },
  {
    "position": 113,
    "kind": {
      "Text": "https://bpaulino"
    },
    "line": 4
  },
  {
    "position": 129,
    "kind": "Dot",
    "line": 4
  },
  {
    "position": 130,
    "kind": {
      "Text": "com"
    },
    "line": 4
  },
  {
    "position": 133,
    "kind": "RightParen",
    "line": 4
  },
  {
    "position": 134,
    "kind": "Newline",
    "line": 5
  },
  {
    "position": 134,
    "kind": "EndOfFile",
    "line": 5
  }
]
